### PR TITLE
Purge and flush

### PR DESCRIPTION
**Pull request by dstuebe**
Changes:
* Add purge method to decrease heap usage when an application is finished writing to the store
* Only flush dirty LookupData
* Adds ObjectStore interface so that purge is available on both CounterStore and AppendStore

### CODE REVIEW

File: src/main/java/com/upserve/uppend/lookup/LongLookup.java [hunk 1]
```diff
@@ -114,7 +114,7 @@ protected boolean removeEldestEntry(Map.Entry<Path, LookupData> eldest) {
      * visible
      *
      * @param partition the partition to look up
-     * @param key the key to look up
+     * @param key       the key to look up
```
Comment (bfulton) [inline src/main/java/com/upserve/uppend/lookup/LongLookup.java]: Extra ws

File: src/main/java/com/upserve/uppend/lookup/LongLookup.java [hunk 2]
```diff
@@ -145,7 +145,7 @@ public long get(String partition, String key) {
      * values are not visible
      *
      * @param partition the partition to look up
-     * @param key the key to look up
+     * @param key       the key to look up
```
Comment (bfulton) [inline src/main/java/com/upserve/uppend/lookup/LongLookup.java]: Extra ws

File: src/main/java/com/upserve/uppend/lookup/LongLookup.java [hunk 3]
```diff
@@ -253,17 +253,21 @@ public void flush() {
...
-                try {
-                    log.trace("cache flushing {}", path);
-                    data.flush();
-                    log.trace("cache flushed {}", path);
-                } catch (Exception e) {
-                    log.error("unable to flush " + path, e);
+        cacheCopy.forEach(
+                (path, data) ->
+                {
+                    if (data.isDirty())
```
Comment (bfulton) [inline src/main/java/com/upserve/uppend/lookup/LongLookup.java]: This seems like something data should decide from within `flush`

File: src/main/java/com/upserve/uppend/FileAppendOnlyStore.java [hunk 4]
```diff
@@ -109,13 +109,27 @@ public void clear() {
     @Override
     protected void flushInternal() {
         // Flush lookups, then blocks, then blobs, since this is the access order of a read.
-        lookups.flush();
+        // Check non null because the super class is registered in the autoflusher before the constructor finishes
```
[the remark this answers is not in the document]
Comment (dstuebe) [reply src/main/java/com/upserve/uppend/FileAppendOnlyStore.java]: Yes - got an NPE. Gross

File: src/main/java/com/upserve/uppend/lookup/LongLookup.java (hunk 3 above)
Comment (dstuebe) [reply src/main/java/com/upserve/uppend/lookup/LongLookup.java]: Seems like a lot of overhead to submit it as a task just to check the whether you need to flush.
Being able to trim (formerly known as purge) the write cache certainly is the best way to reduce the flushes, but the auto flusher pool gets pretty busy.

**Comment by dstuebe** [comment]
Based on what I am seeing in application, I think we may also want to trim the mapped byte buffers in blocked longs in trim. I think they are eating significant resources.

File: src/main/java/com/upserve/uppend/FileAppendOnlyStore.java [hunk 5]
```diff
@@ -109,13 +109,27 @@ public void clear() {
...
+    @Override
+    public void purge() {
+        lookups.close();
         blocks.flush();
         blobs.flush();
     }
 
     @Override
     protected void closeInternal() {
+        // Close blobs first to stop appends
```
[the remark this answers is not in the document]
Comment (dstuebe) [reply src/main/java/com/upserve/uppend/FileAppendOnlyStore.java]: FileStore already has an isClosed - do you want to make that accessible?

**Comment by dstuebe** [comment]
Fixed most of the issues. Would like to get a set of preferences we agree on so I can use reformat code to avoid discussion of white space and syntax entirely.

File: src/main/java/com/upserve/uppend/lookup/LongLookup.java (hunk 2 above)
Comment (dstuebe) [reply src/main/java/com/upserve/uppend/lookup/LongLookup.java]: screen_shot_2018-03-23_at_1_50_01_pm

**Pull request merged**